README.md
# hash-utils
Hash utilities for password encryption


debian/changelog
hash-utils (1.0.0-1) unstable; urgency=medium
  * first version of the package

 -- fpraden <florian@praden.eu>  Thu, 05 Feb 2015 15:11:59 +0100


base64.h
/*
 * Copyright (c) 1996, 1998 by Internet Software Consortium.
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND INTERNET SOFTWARE CONSORTIUM DISCLAIMS
 * ALL WARRANTIES WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL INTERNET SOFTWARE
 * CONSORTIUM BE LIABLE FOR ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL
 * DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR
 * PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS
 * ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS
 * SOFTWARE.
 */

/*
 * Portions Copyright (c) 1995 by International Business Machines, Inc.
 *
 * International Business Machines, Inc. (hereinafter called IBM) grants
 * permission under its copyrights to use, copy, modify, and distribute this
 * Software with or without fee, provided that the above copyright notice and
 * all paragraphs of this notice appear in all copies, and that the name of IBM
 * not be used in connection with the marketing of any product incorporating
 * the Software or modifications thereof, without specific, written prior
 * permission.
 *
 * To the extent it has a right to do so, IBM grants an immunity from suit
 * under its patents, if any, for the use, sale or manufacture of products to
 * the extent that such products are used for performing Domain Name System
 * dynamic updates in TCP/IP networks by means of the Software.  No immunity is
 * granted for any product per se or for any other function of any product.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", AND IBM DISCLAIMS ALL WARRANTIES,
 * INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
 * PARTICULAR PURPOSE.  IN NO EVENT SHALL IBM BE LIABLE FOR ANY SPECIAL,
 * DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER ARISING
 * OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE, EVEN
 * IF IBM IS APPRISED OF THE POSSIBILITY OF SUCH DAMAGES.
 */

#ifndef BASE64_H
#define BASE64_H

#include <stddef.h>
extern size_t base64_encode(char const *src, size_t srclength, char *target, size_t targsize);
extern size_t base64_decode(char const *src, char *target, size_t targsize);

#endif /* BASE64_H */

/* vim:cinoptions=>s,e0,n0,f0,{0,}0,^0,=s,ps,t0,c3,+s,(2s,us,)20,*30,gs,hs
 * vim:ts=8
 * vim:sw=8
 * vim:noexpandtab
 */


passwd-tools.c
// compile with gcc -lcrypt -lcrypto -o passwd-tools passwd-tools.c base64.c
#include <stddef.h>
#include <stdlib.h>
#include <string.h>
#include <stdio.h>

//base64
//#include "base64.h"

#define __USE_GNU
#include <crypt.h>



//md5v2 & sha1 & PBKDF2 -lcrypto
#include <openssl/sha.h>
#include <openssl/evp.h>
#include <openssl/hmac.h>
#include <openssl/pem.h>

#define PBKDF2_SALTLEN		(16)
#define PBKDF2_PRF_DEF		6
#define PBKDF2_ITER_DEF		64000
#define PBKDF2_F_SCAN		":%m[^:]:%u:%m[^:]"
#define PBKDF2_F_PRINT	        ":%s:%lu:%s:%s"

#define EVP_MAX_MDSTR_SIZE 10

const char seedMAP[65]="ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789./";


/* A BASE-64 EN-DE-CODER USING OPENSSL (partially by Len Schulwitz)
 */
/*This function will Base-64 encode your data.*/
unsigned long  base64encode (void *data, int len_data,char *output,unsigned long size)
{
  BIO *b64_bio, *mem_bio; //Declare two BIOs. One base64 encodes, the other stores memory.
  BUF_MEM *mem_bio_mem_ptr; //Pointer to the "memory BIO" structure holding the base64 data.
  unsigned long min_size=0;
  b64_bio = BIO_new(BIO_f_base64()); //Initialize our base64 filter BIO.
  mem_bio = BIO_new(BIO_s_mem()); //Initialize our memory sink BIO.
  BIO_push(b64_bio, mem_bio); //Link the BIOs (i.e. create a filter-sink BIO chain.)
  BIO_set_flags(b64_bio, BIO_FLAGS_BASE64_NO_NL); //Don't add a newline every 64 characters.
  BIO_write(b64_bio, data, len_data); //Encode and write our b64 data.
  BIO_flush(b64_bio); //Flush data. Necessary for b64 encoding, because of pad characters.
  BIO_get_mem_ptr(mem_bio, &mem_bio_mem_ptr); //Store address of mem_bio's memory structure.
  min_size=(size-1 < mem_bio_mem_ptr->length ? size-1:mem_bio_mem_ptr->length);
  memcpy(output,mem_bio_mem_ptr->data,min_size); //Copy memory
  output[min_size]=0;
  BIO_free_all(b64_bio); //Destroys all BIOs in chain, starting with b64 (i.e. the 1st one).
  return min_size;
}
// output allocated space must be >= length
unsigned int base64decode(void *input, int length,unsigned char *output)
{
  BIO *b64_bio, *mem_bio;
  unsigned int len = 0;
  b64_bio = BIO_new(BIO_f_base64());
  mem_bio = BIO_new_mem_buf(input, length);
  mem_bio = BIO_push(b64_bio, mem_bio);
  // BIO_push(b64_bio, mem_bio);
  BIO_set_flags(b64_bio, BIO_FLAGS_BASE64_NO_NL); //Don't add a newline every 64 characters.
  len = BIO_read(mem_bio, output, length);
  // printf("b64_dec_len = %d\n", len);
  output[len] = '\0';
  BIO_free_all(mem_bio);
  return len;
}

void genSalt(unsigned char * salt,unsigned int saltlen,const char *maps,unsigned long lMaps){
  unsigned int seed=0;
  unsigned long n=0;
  seed=time(NULL);
  for(n=0;n<saltlen;n++)
    salt[n]=maps[rand_r(&seed)%lMaps];
  
}
void genSaltBin(unsigned char * salt,unsigned int saltlen){
  unsigned int seed=0;
  unsigned long n=0;
  seed=time(NULL);
  for(n=0;n<saltlen;n++)
    salt[n]=rand_r(&seed);
  
}


char *fDigest(char *str,char *salt,unsigned long lSalt,unsigned char *digest){

  char *ret=NULL;

  EVP_MD_CTX *mdctx;
  const EVP_MD *md;
  unsigned char md_value[EVP_MAX_MD_SIZE];
  char		md_b64[(EVP_MAX_MD_SIZE * 2) + 5];
  char		*salt_b64=NULL,*salt_bin=NULL;

  int md_len,i;
  unsigned long lRet=lSalt+EVP_MAX_MDSTR_SIZE+9+(EVP_MAX_MD_SIZE * 2);

 
  md = EVP_get_digestbyname(digest);

  if(!md) {
    printf("Unknown message digest %s\nUsing sha256\n", digest);
    md = EVP_get_digestbyname("sha256");
    digest="sha256";
  }
  
  ret=(char *)calloc(lRet,sizeof(char));
  
  mdctx = EVP_MD_CTX_create();
  EVP_DigestInit_ex(mdctx, md, NULL);
  
  if(salt){
    if(salt[0]==0){
      genSaltBin(salt,lSalt);
      salt_b64=(char *)calloc(lSalt*2+5,sizeof(char));
      base64_encode((const char *) salt, lSalt,salt_b64, lSalt*2+5);
      //base64encode(salt, lSalt,salt_b64, lSalt*2+5);
    }
    else if(salt[lSalt-1]=='='){
      salt_bin=(char *)calloc(lSalt+1,sizeof(char));
      lSalt=base64_decode((const char *) salt, salt_bin, lSalt+1);
      //lSalt=base64decode( salt, lSalt,salt_bin);
    }
    EVP_DigestUpdate(mdctx, salt_bin?salt_bin:salt, lSalt);
  }
  
  EVP_DigestUpdate(mdctx, str, strlen(str));
  EVP_DigestFinal_ex(mdctx, md_value, &md_len);
  EVP_MD_CTX_destroy(mdctx);
  
  

  memset(md_b64, 0x00, sizeof md_b64);
  base64_encode((const char *) md_value, EVP_MD_size(md),md_b64, sizeof md_b64);
  //base64encode( md_value, EVP_MD_size(md),md_b64, sizeof md_b64);

  snprintf(ret,lRet,":%s:%s:%s",digest,(salt?(salt_b64?salt_b64:salt):""),md_b64);
  if(salt_b64)
    free(salt_b64);
  if(salt_bin)
    free(salt_bin);
  
  return ret;
}


/* This is an implementation of PKCS#5 v2.0 password based encryption key
 * derivation function PBKDF2.
 * SHA1 version verified against test vectors posted by Peter Gutmann
 * <pgut001@cs.auckland.ac.nz> to the PKCS-TNG <pkcs-tng@rsa.com> mailing list.
 */
/* From Atheme */
int PKCS5_PBKDF2_HMAC(const char *pass, int passlen,
			   const unsigned char *salt, int saltlen, int iter,
			   const EVP_MD *digest,
			   int keylen, unsigned char *out)
{
	unsigned char digtmp[EVP_MAX_MD_SIZE], *p=NULL, itmp[4];
	int cplen=0, j=0, k=0, tkeylen=0, mdlen=0;
	unsigned long i = 1,tt;
	HMAC_CTX hctx;

	mdlen = EVP_MD_size(digest);

	HMAC_CTX_init(&hctx);
	p = out;
	tkeylen = keylen;
	if(!pass)
		passlen = 0;
	else if(passlen == -1)
		passlen = strlen(pass);
	while(tkeylen)
	{
		if(tkeylen > mdlen)
			cplen = mdlen;
		else
			cplen = tkeylen;
		/* We are unlikely to ever use more than 256 blocks (5120 bits!)
		 * but just in case...
		 */
		itmp[0] = (unsigned char)((i >> 24) & 0xff);
		itmp[1] = (unsigned char)((i >> 16) & 0xff);
		itmp[2] = (unsigned char)((i >> 8) & 0xff);
		itmp[3] = (unsigned char)(i & 0xff);
				
		HMAC_Init_ex(&hctx, pass, passlen, digest, NULL);
		HMAC_Update(&hctx, salt, saltlen);
		HMAC_Update(&hctx, itmp, 4);
		HMAC_Final(&hctx, digtmp, NULL);
		memcpy(p, digtmp, cplen);
		for(j = 1; j < iter; j++)
		{
			HMAC(digest, pass, passlen,
				 digtmp, mdlen, digtmp, NULL);
			for(k = 0; k < cplen; k++)
				p[k] ^= digtmp[k];
		}
		tkeylen-= cplen;
		i++;
		p+= cplen;
	}
	HMAC_CTX_cleanup(&hctx);
	return 1;
}

/*******************************************************************************************/

char *fPBKDF2(char *str,char *salt,unsigned long lSalt,char *sPRF){

  char *ssalt=NULL,*digestname=NULL,*ssPRF=NULL;
  char		*salt_b64=NULL,*salt_bin=NULL;
  unsigned long iter=24576,prf=0,lRes=0;
  char *result=NULL;

  const EVP_MD*	md = NULL;
  
  unsigned char	digest[EVP_MAX_MD_SIZE];
  char		digest_b64[(EVP_MAX_MD_SIZE * 2) + 5];

  if(salt && (sscanf(salt, PBKDF2_F_SCAN, &ssPRF, &iter, &ssalt) == 3)){
    salt=ssalt;
    sPRF=ssPRF;
    lSalt=strlen(ssalt);
    if(salt[lSalt-1]=='='){
      salt_bin=(char *)calloc(lSalt+1,sizeof(char));
      lSalt=base64_decode((const char *) salt, salt_bin, lSalt+1);
      //lSalt=base64decode( salt, lSalt,salt_bin);
    }
  }
  else{
    if(salt){
      if(salt[0]==0){
	genSaltBin(salt,lSalt);
	salt_b64=(char *)calloc(lSalt*2+5,sizeof(char));
	base64_encode((const char *) salt, lSalt,salt_b64, lSalt*2+5);
	//base64encode( salt, lSalt,salt_b64, lSalt*2+5);
      }
      else if(salt[lSalt-1]=='='){
	salt_bin=(char *)calloc(lSalt+1,sizeof(char));
	lSalt=base64_decode((const char *) salt, salt_bin, lSalt+1);
	//lSalt=base64decode( salt, lSalt,salt_bin);
      }
    }
    else{
      if(lSalt == 0)
	lSalt=PBKDF2_SALTLEN;
    
      ssalt=(char *)calloc(lSalt+1,sizeof(char));
      salt=ssalt;
      genSaltBin(salt,lSalt);
      salt_b64=(char *)calloc(lSalt*2+5,sizeof(char));
      base64_encode((const char *) salt, lSalt,salt_b64, lSalt*2+5);
      //base64encode( salt, lSalt,salt_b64, lSalt*2+5);
    }
  }

  if(iter == 0)
    iter=PBKDF2_ITER_DEF;

  if(sPRF){
    if(strncmp("pbkdf2-",sPRF,7) != 0){
      if(ssPRF)
	free(ssPRF);
      if(ssalt)
	free(ssalt);
      if(salt_b64)
	free(salt_b64);
      if(salt_bin)
	free(salt_bin);
      return NULL;
    }
    digestname=sPRF+7;
  }
  
  md = EVP_get_digestbyname(digestname);
  
  if(!md) {
    printf("Unknown message digest %s\nUsing sha256\n", digest);
    md = EVP_get_digestbyname("sha256");
    digestname="sha256";
  }

  
  /* Compute the PBKDF2 digest */
  size_t sl = strlen(str);
  (void) PKCS5_PBKDF2_HMAC(str, sl, (unsigned char *) salt_bin?salt_bin:salt, lSalt,
	                         iter, md, EVP_MD_size(md), digest);

  /* Convert the digest to Base 64 */
  memset(digest_b64, 0x00, sizeof digest_b64);
  base64_encode((const char *) digest, EVP_MD_size(md), digest_b64, sizeof digest_b64);
  //base64encode( digest, EVP_MD_size(md), digest_b64, sizeof digest_b64);

  /* Format the result */
  lRes=(EVP_MAX_MD_SIZE * 2) + strlen(sPRF)+lSalt+30;
  result=(char *)calloc(lRes,sizeof(char));
  
    
  snprintf(result, lRes, PBKDF2_F_PRINT, sPRF, iter, (salt_b64?salt_b64:salt), digest_b64);

  if(ssPRF)
    free(ssPRF);
  if(ssalt)
    free(ssalt);
  if(salt_b64)
    free(salt_b64);
  if(salt_bin)
    free(salt_bin);
  
  return result;
  
}



char *fposix(char *str,char *salt){
  
  char *ret=NULL,*result=NULL;

  struct crypt_data data;
  char salt2[3];
  unsigned long len=0;
  unsigned int seed=0;

  seed=time(NULL);
  memset(&data,0,sizeof(struct crypt_data));
  
  if(!salt){
    //use default crypt/salt
    salt=salt2;
    salt2[2]=0;
    salt[0]=seedMAP[rand_r(&seed)%64];
    salt[1]=seedMAP[rand_r(&seed)%64];
  }
    
  result=crypt_r(str, salt,&data);
  if(!result){
    //TODO error
    return NULL;
  }
  
  len=strlen(result)+1;
  ret=(char *)calloc(len,sizeof(char));
  strncpy(ret,result,len);
  
  return ret;
}



int main(int argc,char **argv){

  OpenSSL_add_all_digests();

  char *salt=NULL;
  char nsalt[17]={0};
  char lSalt=0;
  char *res=NULL;
  char *digestname;
  
  if(argc< 3)
    return 1;


  if(strcmp(argv[1],"posix") == 0){
    if(argc==3)
      res=fposix(argv[2],NULL);
    else
      res=fposix(argv[2],argv[3]);
  }
  else if(strncmp(argv[1],"pbkdf2-",7) == 0){
    if(argc > 3){
      salt=argv[3];
      lSalt=strlen(salt);
      if(salt[0]==' ')
	salt[0]=0;
    }
    res=fPBKDF2(argv[2],salt,lSalt,argv[1]);
  }
  else{
    if(argc > 3){
      salt=argv[3];
      lSalt=strlen(salt);
      if(salt[0]==' ')
	salt[0]=0;
    }
    res=fDigest(argv[2],salt,lSalt,argv[1]);
  }
  EVP_cleanup();
  if(res){
    printf("%s",res);
    free(res);
    return 0;
  }
  return 1;
}



base64.c
/*
 * Copyright (c) 1996, 1998 by Internet Software Consortium.
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND INTERNET SOFTWARE CONSORTIUM DISCLAIMS
 * ALL WARRANTIES WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL INTERNET SOFTWARE
 * CONSORTIUM BE LIABLE FOR ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL
 * DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR
 * PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS
 * ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS
 * SOFTWARE.
 */

/*
 * Portions Copyright (c) 1995 by International Business Machines, Inc.
 *
 * International Business Machines, Inc. (hereinafter called IBM) grants
 * permission under its copyrights to use, copy, modify, and distribute this
 * Software with or without fee, provided that the above copyright notice and
 * all paragraphs of this notice appear in all copies, and that the name of IBM
 * not be used in connection with the marketing of any product incorporating
 * the Software or modifications thereof, without specific, written prior
 * permission.
 *
 * To the extent it has a right to do so, IBM grants an immunity from suit
 * under its patents, if any, for the use, sale or manufacture of products to
 * the extent that such products are used for performing Domain Name System
 * dynamic updates in TCP/IP networks by means of the Software.  No immunity is
 * granted for any product per se or for any other function of any product.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", AND IBM DISCLAIMS ALL WARRANTIES,
 * INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
 * PARTICULAR PURPOSE.  IN NO EVENT SHALL IBM BE LIABLE FOR ANY SPECIAL,
 * DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER ARISING
 * OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE, EVEN
 * IF IBM IS APPRISED OF THE POSSIBILITY OF SUCH DAMAGES.
 */

#include "base64.h"
#include <string.h>

static const char Base64[] =
	"ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
static const char Pad64 = '=';

/* (From RFC1521 and draft-ietf-dnssec-secext-03.txt)
   The following encoding technique is taken from RFC 1521 by Borenstein
   and Freed.  It is reproduced here in a slightly edited form for
   convenience.

   A 65-character subset of US-ASCII is used, enabling 6 bits to be
   represented per printable character. (The extra 65th character, "=",
   is used to signify a special processing function.)

   The encoding process represents 24-bit groups of input bits as output
   strings of 4 encoded characters. Proceeding from left to right, a
   24-bit input group is formed by concatenating 3 8-bit input groups.
   These 24 bits are then treated as 4 concatenated 6-bit groups, each
   of which is translated into a single digit in the base64 alphabet.

   Each 6-bit group is used as an index into an array of 64 printable
   characters. The character referenced by the index is placed in the
   output string.

                         Table 1: The Base64 Alphabet

      Value Encoding  Value Encoding  Value Encoding  Value Encoding
          0 A            17 R            34 i            51 z
          1 B            18 S            35 j            52 0
          2 C            19 T            36 k            53 1
          3 D            20 U            37 l            54 2
          4 E            21 V            38 m            55 3
          5 F            22 W            39 n            56 4
          6 G            23 X            40 o            57 5
          7 H            24 Y            41 p            58 6
          8 I            25 Z            42 q            59 7
          9 J            26 a            43 r            60 8
         10 K            27 b            44 s            61 9
         11 L            28 c            45 t            62 +
         12 M            29 d            46 u            63 /
         13 N            30 e            47 v
         14 O            31 f            48 w         (pad) =
         15 P            32 g            49 x
         16 Q            33 h            50 y

   Special processing is performed if fewer than 24 bits are available
   at the end of the data being encoded.  A full encoding quantum is
   always completed at the end of a quantity.  When fewer than 24 input
   bits are available in an input group, zero bits are added (on the
   right) to form an integral number of 6-bit groups.  Padding at the
   end of the data is performed using the '=' character.

   Since all base64 input is an integral number of octets, only the
         -------------------------------------------------
   following cases can arise:

       (1) the final quantum of encoding input is an integral
           multiple of 24 bits; here, the final unit of encoded
	   output will be an integral multiple of 4 characters
	   with no "=" padding,
       (2) the final quantum of encoding input is exactly 8 bits;
           here, the final unit of encoded output will be two
	   characters followed by two "=" padding characters, or
       (3) the final quantum of encoding input is exactly 16 bits;
           here, the final unit of encoded output will be three
	   characters followed by one "=" padding character.
   */

size_t base64_encode(char const *src, size_t srclength, char *target, size_t targsize)
{
	size_t datalength = 0;
	unsigned char input[3];
	unsigned char output[4];
	size_t i;

	while (2 < srclength) {
		input[0] = *src++;
		input[1] = *src++;
		input[2] = *src++;
		srclength -= 3;

		output[0] = input[0] >> 2;
		output[1] = ((input[0] & 0x03) << 4) + (input[1] >> 4);
		output[2] = ((input[1] & 0x0f) << 2) + (input[2] >> 6);
		output[3] = input[2] & 0x3f;
		/*		
		soft_assert(output[0] < 64);
		soft_assert(output[1] < 64);
		soft_assert(output[2] < 64);
		soft_assert(output[3] < 64);
		*/

		if (datalength + 4 > targsize)
			return (-1);
		target[datalength++] = Base64[output[0]];
		target[datalength++] = Base64[output[1]];
		target[datalength++] = Base64[output[2]];
		target[datalength++] = Base64[output[3]];
	}

	/* Now we worry about padding. */
	if (0 != srclength) {
		/* Get what's left. */
		input[0] = input[1] = input[2] = '\0';
		for (i = 0; i < srclength; i++)
			input[i] = *src++;

		output[0] = input[0] >> 2;
		output[1] = ((input[0] & 0x03) << 4) + (input[1] >> 4);
		output[2] = ((input[1] & 0x0f) << 2) + (input[2] >> 6);
		/*
		soft_assert(output[0] < 64);
		soft_assert(output[1] < 64);
		soft_assert(output[2] < 64);
		*/
		if (datalength + 4 > targsize)
			return (-1);
		target[datalength++] = Base64[output[0]];
		target[datalength++] = Base64[output[1]];
		if (srclength == 1)
			target[datalength++] = Pad64;
		else
			target[datalength++] = Base64[output[2]];
		target[datalength++] = Pad64;
	}
	if (datalength >= targsize)
		return (-1);
	target[datalength] = '\0';	/* Returned value doesn't count \0. */
	return (datalength);
}

/* skips all whitespace anywhere.
   converts characters, four at a time, starting at (or after)
   src from base - 64 numbers into three 8 bit bytes in the target area.
   it returns the number of data bytes stored at the target, or -1 on error.
 */
size_t base64_decode(char const *src, char *target, size_t targsize)
{
	int tarindex, state, ch;
	char *pos;

	state = 0;
	tarindex = 0;

	while ((ch = *src++) != '\0') {
		if (isspace((unsigned char)ch))        /* Skip whitespace anywhere. */
			continue;

		if (ch == Pad64)
			break;

		pos = strchr(Base64, ch);
		if (pos == 0) 		/* A non-base64 character. */
			return (-1);

		switch (state) {
		case 0:
			if (target) {
				if ((size_t)tarindex >= targsize)
					return (-1);
				target[tarindex] = (pos - Base64) << 2;
			}
			state = 1;
			break;
		case 1:
			if (target) {
				if ((size_t)tarindex + 1 >= targsize)
					return (-1);
				target[tarindex]   |=  (pos - Base64) >> 4;
				target[tarindex+1]  = ((pos - Base64) & 0x0f)
							<< 4 ;
			}
			tarindex++;
			state = 2;
			break;
		case 2:
			if (target) {
				if ((size_t)tarindex + 1 >= targsize)
					return (-1);
				target[tarindex]   |=  (pos - Base64) >> 2;
				target[tarindex+1]  = ((pos - Base64) & 0x03)
							<< 6;
			}
			tarindex++;
			state = 3;
			break;
		case 3:
			if (target) {
				if ((size_t)tarindex >= targsize)
					return (-1);
				target[tarindex] |= (pos - Base64);
			}
			tarindex++;
			state = 0;
			break;
		default:
			return 0;
		}
	}

	/*
	 * We are done decoding Base-64 chars.  Let's see if we ended
	 * on a byte boundary, and/or with erroneous trailing characters.
	 */

	if (ch == Pad64) {		/* We got a pad char. */
		ch = *src++;		/* Skip it, get next. */
		switch (state) {
		case 0:		/* Invalid = in first position */
		case 1:		/* Invalid = in second position */
			return (-1);

		case 2:		/* Valid, means one byte of info */
			/* Skip any number of spaces. */
			for ((void)NULL; ch != '\0'; ch = *src++)
				if (!isspace((unsigned char)ch))
					break;
			/* Make sure there is another trailing = sign. */
			if (ch != Pad64)
				return (-1);
			ch = *src++;		/* Skip the = */
			/* Fall through to "single trailing =" case. */
			/* FALLTHROUGH */

		case 3:		/* Valid, means two bytes of info */
			/*
			 * We know this char is an =.  Is there anything but
			 * whitespace after it?
			 */
			for ((void)NULL; ch != '\0'; ch = *src++)
				if (!isspace((unsigned char)ch))
					return (-1);

			/*
			 * Now make sure for cases 2 and 3 that the "extra"
			 * bits that slopped past the last full byte were
			 * zeros.  If we don't check them, they become a
			 * subliminal channel.
			 */
			if (target && target[tarindex] != 0)
				return (-1);
		}
	} else {
		/*
		 * We ended by seeing the end of the string.  Make sure we
		 * have no partial bytes lying around.
		 */
		if (state != 0)
			return (-1);
	}

	return (tarindex);
}
